src/dune_rules/module_compilation.ml

```diff
@@ -218,6 +219,50 @@ let build_cm cctx ~precompiled_cmi ~cm_kind (m : Module.t)
     >>| Action.Full.add_sandbox sandbox))
   |> Memo.Option.iter ~f:Fun.id
 
+let build_melange_js ~cctx m =
+  let cm_kind = Lib_mode.Cm_kind.Melange Cmj in
+  let sctx = CC.super_context cctx in
+  let dir = CC.dir cctx in
+  let obj_dir = CC.obj_dir cctx in
+  let ctx = Super_context.context sctx in
+  let mode = Lib_mode.of_cm_kind cm_kind in
+  let ml_kind = Lib_mode.Cm_kind.source cm_kind in
+  (let open Option.O in
+  let+ compiler = Result.to_option (Context.compiler ctx mode) in
+  let src = Obj_dir.Module.cm_file_exn obj_dir m ~kind:cm_kind in
+  let output =
+    Obj_dir.Module.obj_file obj_dir m ~kind:cm_kind ~ext:Melange.js_ext
+  in
+  let obj_dirs =
+    Obj_dir.all_obj_dirs obj_dir ~mode
+    |> List.concat_map ~f:(fun p ->
+           [ Command.Args.A "-I"; Path (Path.build p) ])
+  in
+  let dep_graph = Ml_kind.Dict.get (CC.dep_graphs cctx) ml_kind in
+  let cmj_deps =
+    Action_builder.dyn_paths_unit
+      (let open Action_builder.O in
+      let+ deps = Dep_graph.deps_of dep_graph m in
+      List.concat_map deps ~f:(fun m ->
+          if Module.has m ~ml_kind:Impl && cm_kind = Melange Cmj then
+            [ Path.build
+                (Obj_dir.Module.obj_file obj_dir m ~kind:cm_kind
+                   ~ext:Melange.js_ext)
+            ]
+          else []))
+  in
+  Super_context.add_rule sctx ~dir ?loc:(CC.loc cctx)
+    (let open Action_builder.With_targets.O in
+    Action_builder.with_no_targets cmj_deps
+    >>> Command.run ~dir:(Path.build dir) (Ok compiler)
+          [ Command.Args.S obj_dirs
+          ; Command.Args.as_any (CC.melange_js_includes cctx)
+          ; A "-o"
+          ; Target output
+          ; Dep (Path.build src)
+          ]))
```

Comment:
This is roughly the same as how cmj rules for melange are created, but:
- removing unnecessary deps like cmis
- not using `-impl` flag for the target
- removed `-no-alias-deps` and a lot of other flags
- removed the `-open` flags that are used when ` Modules.alias_for`.

cc @anmonteiro in case I'm forgetting something, I'm not sure if separate `.js` generation is currently used in `dune.mel` pipeline.